File: tests/pages/payroll_page.py
"""
Payroll Workspace Page Object (/payroll).
"""

from playwright.sync_api import Page, expect
from tests.pages.base_page import BasePage
from tests.components.modal import ModalComponent


class PayrollPage(BasePage):
    PATH = "/payroll"

    def __init__(self, page: Page):
        super().__init__(page)
        self.stepper = page.locator("[data-testid='stepper'], ol, .stepper, nav[aria-label='Progress']").first
        self.run_payroll_button = (
            page.locator("[data-testid='run-payroll-btn'], button")
            .filter(has_text="Execute Bulk Payroll Run")
            .or_(page.locator("[data-testid='reopen-payroll-btn']"))
            .or_(page.locator("button").filter(has_text="Run Payroll"))
            .first
        )
        self.finalize_button = (
            page.locator("[data-testid='finalize-payroll-btn'], button")
            .filter(has_text="Finalize Payroll")
            .or_(page.locator("button").filter(has_text="Finalize"))
            .first
        )
        self.publish_button = (
            page.locator("[data-testid='publish-payroll-btn'], button")
            .filter(has_text="Publish Payslips")
            .or_(page.locator("button").filter(has_text="Publish"))
            .first
        )
        self.reopen_button = page.locator("[data-testid='reopen-payroll-btn'], button").filter(has_text="Reopen").first
        self.confirm_dialog = ModalComponent(page, test_id="confirm-dialog")
        self.download_payslip_btn = (
            page.locator("[data-testid='download-payslip-btn'], button, a")
            .filter(has_text="Download Payslip")
            .or_(page.locator("button, a").filter(has_text="PDF"))
            .first
        )

    def click_run_payroll(self):
        """Initiates payroll calculation run."""
        expect(self.run_payroll_button).to_be_visible()
        self.run_payroll_button.click()

    def click_finalize(self):
        """Finalizes the calculated payroll cycle."""
        expect(self.finalize_button).to_be_visible()
        self.finalize_button.click()

    def click_publish(self):
        """Publishes finalized payslips to employees."""
        expect(self.publish_button).to_be_visible()
        self.publish_button.click()

    def reopen_payroll_for_revision(self):
        """Reopens payroll cycle via confirmation dialog."""
        expect(self.reopen_button).to_be_visible()
        self.reopen_button.click()
        if self.confirm_dialog.is_open():
            self.confirm_dialog.confirm()

File: e2e/pages/PayrollPage.ts
import { Page, Locator, expect } from "@playwright/test";
import { BasePage } from "./BasePage";

export class PayrollPage extends BasePage {
  readonly header: Locator;
  readonly stepper: Locator;
  readonly runPayrollButton: Locator;
  readonly finalizePayrollButton: Locator;
  readonly reopenPayrollButton: Locator;
  readonly viewPayslipButtons: Locator;

  constructor(page: Page, baseURL?: string) {
    super(page, baseURL);
    this.header = page.locator('[data-testid="payroll-header"]');
    this.stepper = page.locator('[data-testid="stepper"], [data-testid="payroll-stepper"]');
    this.runPayrollButton = page.locator('[data-testid="run-payroll-btn"]');
    this.finalizePayrollButton = page.locator('[data-testid="finalize-payroll-btn"]');
    this.reopenPayrollButton = page.locator('[data-testid="reopen-payroll-btn"]');
    this.viewPayslipButtons = page.locator('[data-testid="view-payslip-btn"]');
  }

  async goto(): Promise<void> {
    await this.navigate("/payroll");
  }

  async assertLoaded(): Promise<void> {
    await expect(this.page).toHaveURL(/.*\/payroll.*/);
    await expect(this.page.locator("body")).toContainText(/Payroll/i);
  }

  async runPayrollCycle(): Promise<void> {
    if (await this.runPayrollButton.isVisible()) {
      await this.runPayrollButton.click();
    }
  }

  async finalizePayroll(): Promise<void> {
    if (await this.finalizePayrollButton.isVisible()) {
      await this.finalizePayrollButton.click();
    }
  }
}
